File: libs/ai-kernel/src/models/provider.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use sqlx::{FromRow, types::Json};
use uuid::Uuid;

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
pub struct ProviderRoutingRules {
    #[serde(default)]
    pub use_cases: Vec<String>,
    #[serde(default)]
    pub preferred_regions: Vec<String>,
    #[serde(default)]
    pub fallback_provider_ids: Vec<Uuid>,
    #[serde(default = "default_routing_weight")]
    pub weight: i32,
    #[serde(default = "default_context_tokens")]
    pub max_context_tokens: i32,
    #[serde(default = "default_network_scope")]
    pub network_scope: String,
    #[serde(default = "default_supported_modalities")]
    pub supported_modalities: Vec<String>,
    #[serde(default)]
    pub input_cost_per_1k_tokens_usd: f32,
    #[serde(default)]
    pub output_cost_per_1k_tokens_usd: f32,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ProviderHealthState {
    pub status: String,
    pub avg_latency_ms: i32,
    pub error_rate: f32,
    pub last_checked_at: DateTime<Utc>,
}

impl Default for ProviderHealthState {
    fn default() -> Self {
        Self {
            status: "healthy".to_string(),
            avg_latency_ms: 620,
            error_rate: 0.01,
            last_checked_at: Utc::now(),
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct LlmProvider {
    pub id: Uuid,
    pub name: String,
    pub provider_type: String,
    pub model_name: String,
    pub endpoint_url: String,
    pub api_mode: String,
    pub credential_reference: Option<String>,
    pub credential_configured: bool,
    pub enabled: bool,
    pub load_balance_weight: i32,
    pub max_output_tokens: i32,
    pub cost_tier: String,
    pub tags: Vec<String>,
    pub route_rules: ProviderRoutingRules,
    pub health_state: ProviderHealthState,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ListProvidersResponse {
    pub data: Vec<LlmProvider>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CreateProviderRequest {
    pub name: String,
    #[serde(default = "default_provider_type")]
    pub provider_type: String,
    #[serde(default = "default_model_name")]
    pub model_name: String,
    #[serde(default = "default_endpoint_url")]
    pub endpoint_url: String,
    #[serde(default = "default_api_mode")]
    pub api_mode: String,
    #[serde(default)]
    pub credential_reference: Option<String>,
    #[serde(default = "default_enabled")]
    pub enabled: bool,
    #[serde(default = "default_provider_weight")]
    pub load_balance_weight: i32,
    #[serde(default = "default_output_tokens")]
    pub max_output_tokens: i32,
    #[serde(default = "default_cost_tier")]
    pub cost_tier: String,
    #[serde(default)]
    pub tags: Vec<String>,
    #[serde(default)]
    pub route_rules: Option<ProviderRoutingRules>,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
pub struct UpdateProviderRequest {
    pub name: Option<String>,
    pub provider_type: Option<String>,
    pub model_name: Option<String>,
    pub endpoint_url: Option<String>,
    pub api_mode: Option<String>,
    pub credential_reference: Option<String>,
    pub enabled: Option<bool>,
    pub load_balance_weight: Option<i32>,
    pub max_output_tokens: Option<i32>,
    pub cost_tier: Option<String>,
    pub tags: Option<Vec<String>>,
    pub route_rules: Option<ProviderRoutingRules>,
    pub health_state: Option<ProviderHealthState>,
}

#[derive(Debug, Clone, FromRow)]
pub(crate) struct ProviderRow {
    pub id: Uuid,
    pub name: String,
    pub provider_type: String,
    pub model_name: String,
    pub endpoint_url: String,
    pub api_mode: String,
    pub credential_reference: Option<String>,
    pub enabled: bool,
    pub load_balance_weight: i32,
    pub max_output_tokens: i32,
    pub cost_tier: String,
    pub tags: Json<Vec<String>>,
    pub route_rules: Json<ProviderRoutingRules>,
    pub health_state: Json<ProviderHealthState>,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

impl From<ProviderRow> for LlmProvider {
    fn from(value: ProviderRow) -> Self {
        let credential_configured = value
            .credential_reference
            .as_ref()
            .map(|reference| !reference.trim().is_empty())
            .unwrap_or(false);

        Self {
            id: value.id,
            name: value.name,
            provider_type: value.provider_type,
            model_name: value.model_name,
            endpoint_url: value.endpoint_url,
            api_mode: value.api_mode,
            credential_reference: value.credential_reference,
            credential_configured,
            enabled: value.enabled,
            load_balance_weight: value.load_balance_weight,
            max_output_tokens: value.max_output_tokens,
            cost_tier: value.cost_tier,
            tags: value.tags.0,
            route_rules: value.route_rules.0,
            health_state: value.health_state.0,
            created_at: value.created_at,
            updated_at: value.updated_at,
        }
    }
}

fn default_routing_weight() -> i32 {
    100
}

fn default_context_tokens() -> i32 {
    32_000
}

fn default_provider_type() -> String {
    "openai".to_string()
}

fn default_network_scope() -> String {
    "public".to_string()
}

fn default_supported_modalities() -> Vec<String> {
    vec!["text".to_string()]
}

fn default_model_name() -> String {
    "gpt-4.1-mini".to_string()
}

fn default_endpoint_url() -> String {
    "https://api.openai.com/v1".to_string()
}

fn default_api_mode() -> String {
    "chat_completions".to_string()
}

fn default_enabled() -> bool {
    true
}

fn default_provider_weight() -> i32 {
    100
}

fn default_output_tokens() -> i32 {
    2048
}

fn default_cost_tier() -> String {
    "standard".to_string()
}

File: openfoundry-go/libs/ai-kernel-go/models/provider.go
package models

import (
	"encoding/json"
	"time"

	"github.com/google/uuid"
)

// ProviderRoutingRules controls how the LLM gateway picks a provider.
type ProviderRoutingRules struct {
	UseCases                 []string    `json:"use_cases"`
	PreferredRegions         []string    `json:"preferred_regions"`
	FallbackProviderIDs      []uuid.UUID `json:"fallback_provider_ids"`
	Weight                   int32       `json:"weight"`
	MaxContextTokens         int32       `json:"max_context_tokens"`
	NetworkScope             string      `json:"network_scope"`
	SupportedModalities      []string    `json:"supported_modalities"`
	InputCostPer1KTokensUSD  float32     `json:"input_cost_per_1k_tokens_usd"`
	OutputCostPer1KTokensUSD float32     `json:"output_cost_per_1k_tokens_usd"`
}

// DefaultProviderRoutingRules mirrors the Rust impl Default for
// ProviderRoutingRules + serde defaults: weight=100,
// max_context_tokens=32_000, network_scope="public",
// supported_modalities=["text"].
func DefaultProviderRoutingRules() ProviderRoutingRules {
	return ProviderRoutingRules{
		Weight:              100,
		MaxContextTokens:    32_000,
		NetworkScope:        "public",
		SupportedModalities: []string{"text"},
	}
}

// ProviderHealthState is the rolling health probe summary.
type ProviderHealthState struct {
	Status        string    `json:"status"`
	AvgLatencyMs  int32     `json:"avg_latency_ms"`
	ErrorRate     float32   `json:"error_rate"`
	LastCheckedAt time.Time `json:"last_checked_at"`
}

// LlmProvider is the catalog row for one LLM endpoint.
type LlmProvider struct {
	ID                   uuid.UUID            `json:"id"`
	Name                 string               `json:"name"`
	ProviderType         string               `json:"provider_type"`
	ModelName            string               `json:"model_name"`
	EndpointURL          string               `json:"endpoint_url"`
	APIMode              string               `json:"api_mode"`
	CredentialReference  *string              `json:"credential_reference"`
	CredentialConfigured bool                 `json:"credential_configured"`
	Enabled              bool                 `json:"enabled"`
	LoadBalanceWeight    int32                `json:"load_balance_weight"`
	MaxOutputTokens      int32                `json:"max_output_tokens"`
	CostTier             string               `json:"cost_tier"`
	Tags                 []string             `json:"tags"`
	RouteRules           ProviderRoutingRules `json:"route_rules"`
	HealthState          ProviderHealthState  `json:"health_state"`
	CreatedAt            time.Time            `json:"created_at"`
	UpdatedAt            time.Time            `json:"updated_at"`
}

// ListProvidersResponse is the {"data": [...]} list envelope.
type ListProvidersResponse struct {
	Data []LlmProvider `json:"data"`
}

// CreateProviderRequest defaults match Rust serde:
// provider_type="openai", model_name="gpt-4.1-mini",
// endpoint_url="https://api.openai.com/v1", api_mode="chat_completions",
// enabled=true, load_balance_weight=100, max_output_tokens=2048,
// cost_tier="standard".
type CreateProviderRequest struct {
	Name                string                `json:"name"`
	ProviderType        *string               `json:"provider_type"`
	ModelName           *string               `json:"model_name"`
	EndpointURL         *string               `json:"endpoint_url"`
	APIMode             *string               `json:"api_mode"`
	CredentialReference *string               `json:"credential_reference"`
	Enabled             *bool                 `json:"enabled"`
	LoadBalanceWeight   *int32                `json:"load_balance_weight"`
	MaxOutputTokens     *int32                `json:"max_output_tokens"`
	CostTier            *string               `json:"cost_tier"`
	Tags                []string              `json:"tags"`
	RouteRules          *ProviderRoutingRules `json:"route_rules"`
}

type UpdateProviderRequest struct {
	Name                *string               `json:"name"`
	ProviderType        *string               `json:"provider_type"`
	ModelName           *string               `json:"model_name"`
	EndpointURL         *string               `json:"endpoint_url"`
	APIMode             *string               `json:"api_mode"`
	CredentialReference *string               `json:"credential_reference"`
	Enabled             *bool                 `json:"enabled"`
	LoadBalanceWeight   *int32                `json:"load_balance_weight"`
	MaxOutputTokens     *int32                `json:"max_output_tokens"`
	CostTier            *string               `json:"cost_tier"`
	Tags                *[]string             `json:"tags"`
	RouteRules          *ProviderRoutingRules `json:"route_rules"`
	HealthState         *ProviderHealthState  `json:"health_state"`
}

// Provider creation defaults — exposed for the HTTP-handler slice.
const (
	DefaultProviderType            = "openai"
	DefaultModelName               = "gpt-4.1-mini"
	DefaultEndpointURL             = "https://api.openai.com/v1"
	DefaultAPIMode                 = "chat_completions"
	DefaultCostTier                = "standard"
	DefaultLoadBalanceWeight int32 = 100
	DefaultMaxOutputTokens   int32 = 2048
)

func (r *ProviderRoutingRules) UnmarshalJSON(data []byte) error {
	type alias ProviderRoutingRules
	defaults := DefaultProviderRoutingRules()
	*r = defaults
	return json.Unmarshal(data, (*alias)(r))
}

func DefaultProviderHealthState(now time.Time) ProviderHealthState {
	return ProviderHealthState{Status: "healthy", AvgLatencyMs: 620, ErrorRate: 0.01, LastCheckedAt: now}
}

func (r *CreateProviderRequest) UnmarshalJSON(data []byte) error {
	type alias CreateProviderRequest
	aux := struct{ *alias }{alias: (*alias)(r)}
	if err := json.Unmarshal(data, &aux); err != nil {
		return err
	}
	if r.ProviderType == nil {
		r.ProviderType = ptrOf(DefaultProviderType)
	}
	if r.ModelName == nil {
		r.ModelName = ptrOf(DefaultModelName)
	}
	if r.EndpointURL == nil {
		r.EndpointURL = ptrOf(DefaultEndpointURL)
	}
	if r.APIMode == nil {
		r.APIMode = ptrOf(DefaultAPIMode)
	}
	if r.Enabled == nil {
		r.Enabled = ptrOf(true)
	}
	if r.LoadBalanceWeight == nil {
		r.LoadBalanceWeight = ptrOf(DefaultLoadBalanceWeight)
	}
	if r.MaxOutputTokens == nil {
		r.MaxOutputTokens = ptrOf(DefaultMaxOutputTokens)
	}
	if r.CostTier == nil {
		r.CostTier = ptrOf(DefaultCostTier)
	}
	if r.Tags == nil {
		r.Tags = []string{}
	}
	return nil
}
